src/v/cluster/archival/ntp_archiver_service.cc

```diff
@@ -338,6 +338,15 @@ ss::future<> ntp_archiver::upload_until_abort(bool legacy_mode) {
     }
     if (!_probe) {
         _probe.emplace(_conf->ntp_metrics_disabled, _ntp);
+
+        // Ensure we are exception safe and won't leave the probe without
+        // a watcher in case of exceptions. Also ensures we won't crash calling
+        // the callback.
+        static_assert(noexcept(update_probe()));
+
+        update_probe();
+        _parent.archival_meta_stm()->subscribe_to_state_change(
```

Comment:
Am I missing where the `ntp_archiver_service`'s `_stm_sub_id` is actually assigned per the return value of `subscribe_to_state_change()`?

Reply:
[[nodiscard]] would have really helped :) Appreciate great feedback!